query-algebrizer/src/cc.rs

```diff
@@ -0,0 +1,795 @@
+    /// function deduces that and calls `self.mark_known_empty`. #293.
+    pub fn expand_type_tags(&mut self) {
+        // TODO.
+    }
+}
+
+/// Application.
+impl ConjoiningClauses {
+
+    /// Apply the constraints in the provided pattern to this CC.
+    ///
+    /// This is a single-pass process, which means it is naturally incomplete, failing to take into
+    /// account all information spread across two patterns.
+    ///
+    /// If the constraints cannot be satisfied -- for example, if this pattern includes a numeric
+    /// attribute and a string value -- then the `is_known_empty` field on the CC is flipped and
+    /// the function returns.
+    ///
+    /// A pattern being impossible to satisfy isn't necessarily a bad thing -- this query might
+    /// have branched clauses that apply to different knowledge bases, and might refer to
+    /// vocabulary that isn't (yet) used in this one.
+    ///
+    /// Most of the work done by this function depends on the schema and ident maps in the DB. If
+    /// these change, then any work done is invalid.
+    ///
+    /// There's a lot more we can do here and later by examining the
+    /// attribute:
+    ///
+    /// - If it's unique, and we have patterns like
+    ///
+    ///     [?x :foo/unique 5] [?x :foo/unique ?y]
+    ///
+    ///   then we can either prove impossibility (the two objects must
+    ///   be equal) or deduce identity and simplify the query.
+    ///
+    /// - The same, if it's cardinality-one and the entity is known.
+    ///
+    /// - If there's a value index on this attribute, we might want to
+    ///   run this pattern early in the query.
+    ///
+    /// - A unique-valued attribute can sometimes be rewritten into an
+    ///   existence subquery instead of a join.
+    fn apply_pattern_clause_for_alias<'s>(&mut self, schema: &'s Schema, pattern: &Pattern, alias: &SourceAlias) {
+        if self.is_known_empty {
+            return;
+        }
+
+        // Process each place in turn, applying constraints.
+        // Both `e` and `a` must be entities, which is equivalent here
+        // to being typed as Ref.
+        // Sorry for the duplication; Rust makes it a pain to abstract this.
+
+        // The transaction part of a pattern must be an entid, variable, or placeholder.
+        self.mark_as_tx(&pattern.tx);
+        self.mark_as_ref(&pattern.entity);
+        self.mark_as_ref(&pattern.attribute);
+
+        match pattern.entity {
+            PatternNonValuePlace::Placeholder =>
+                // Placeholders don't contribute any bindings, nor do
+                // they constrain the query -- there's no need to produce
+                // IS NOT NULL, because we don't store nulls in our schema.
+                (),
+            PatternNonValuePlace::Variable(ref v) =>
+                self.bind_column_to_var(alias.1.clone(), DatomsColumn::Entity, v.clone()),
```

Comment:
Can you pull a reference to `alias.1` out of these?  (I'm okay with an unnamed tuple globally but let's help ourselves locally.)